airflow/contrib/hooks/sqoop_hook.py

```diff
@@ -14,210 +14,282 @@
+        self.namenode = connection_parameters.get('namenode', None)
+        self.libjars = connection_parameters.get('libjars', None)
+        self.files = connection_parameters.get('files', None)
+        self.archives = connection_parameters.get('archives', None)
+        self.password_file = connection_parameters.get('password_file', None)
+        self.verbose = verbose
+        self.num_mappers = str(num_mappers)
+        self.properties = properties
 
     def get_conn(self):
         pass
 
-    def Popen(self, cmd, export=False, **kwargs):  # noqa
+    def Popen(self, cmd, **kwargs):
         """
         Remote Popen
 
         :param cmd: command to remotely execute
         :param kwargs: extra arguments to Popen (see subprocess.Popen)
         :return: handle to subprocess
         """
-        prefixed_cmd = self._prepare_command(cmd, export=export)
-        return subprocess.Popen(prefixed_cmd, **kwargs)
-
-    def _prepare_command(self, cmd, export=False):
+        process = subprocess.Popen(cmd,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE,
+                                   **kwargs)
+        output, stderr = process.communicate()
+
+        if process.returncode != 0:
+            raise AirflowException((
+                                       "Cannot execute {} on {}. Error code is: {}"
+                                       "Output: {}, Stderr: {}"
+                                   ).format(cmd, self.conn.host,
+                                            process.returncode, output,
+                                            stderr))
+
+    def _prepare_command(self, export=False):
         if export:
-            connection_cmd = ["sqoop", "export", "--verbose"]
+            connection_cmd = ["sqoop", "export"]
         else:
-            connection_cmd = ["sqoop", "import", "--verbose"]
+            connection_cmd = ["sqoop", "import"]
 
+        if self.verbose:
+            connection_cmd += ["--verbose"]
         if self.job_tracker:
             connection_cmd += ["-jt", self.job_tracker]
         if self.conn.login:
             connection_cmd += ["--username", self.conn.login]
-        # todo: put this in a password file
         if self.conn.password:
             connection_cmd += ["--password", self.conn.password]
-        if self.lib_jars:
-            connection_cmd += ["-libjars", self.lib_jars]
+        if self.password_file:
+            connection_cmd += ["--password-file", self.password_file]
+        if self.libjars:
+            connection_cmd += ["-libjars", self.libjars]
         if self.files:
             connection_cmd += ["-files", self.files]
         if self.namenode:
             connection_cmd += ["-fs", self.namenode]
         if self.archives:
             connection_cmd += ["-archives", self.archives]
+        if self.num_mappers:
+            connection_cmd += ["--num-mappers", self.num_mappers]
```

Comment:
could make the str cast here

Reply:
you fixed this in [AIRFLOW-1217](https://github.com/apache/incubator-airflow/commit/4f459b64e098f4fd73af23030e7992e0465bfe21) :).